src/backend/game_logic.js

```diff
@@ -0,0 +1,53 @@
+import { firestore, root } from '../config/firebase';
+
+export const all_chars = ["Duke", "Assassin", "Contessa", "Captain", "Ambassador"];
+
+const num_of_each_card = 3;
+
+export function distributeCards(roomName) {
+	// helper functions
+	function shuffle(a) {
+		for (let i = a.length - 1; i > 0; i--) {
+			const j = Math.floor(Math.random() * (i + 1));
+			[a[i], a[j]] = [a[j], a[i]];
+		}
+		return a;
+	}
+
+	let cards = [];
+	all_chars.forEach(char => {
+		for (let i = 0; i < num_of_each_card; ++i) {
+			cards.push(char);
+		}
+	});
+	shuffle(cards);
+	
+	const roomRef = firestore.collection(root).doc(roomName);
+	const playerCollection = roomRef.collection("players");
+	playerCollection.get().then((players) => {
+		// give each player 2 cards and 2 coins
+		players.forEach(function(doc) {
+			let playerCards = [cards.pop(), cards.pop()];
+			playerCollection.doc(doc.id).update({cards: playerCards, coins: 2})
+				.then(r => console.log("Successfully distributed cards"))
+				.catch(e => console.log(e));
+		});
+		// put remaining cards in room
+		roomRef.update({cards: cards});
+	}).catch((error) => {
+		// TODO: report error to user
+		console.log("Error getting document:", error);
+	});
+}
+
+export function playerStateCallback(roomName, playerId) {
```

Comment:
For whoever is going to work on the UI, use this callback to update what's shown to the player. `doc.data` will have the number of coins and cards that the player has in possession.